pennylane_forest/qvm.py

```diff
@@ -228,14 +236,19 @@ def apply_parametric_program(self, operations, **kwargs):
         self.prog += self.apply_rotations(rotations)
 
     def generate_samples(self):
+        if self.parametric_compilation:
+            for region, value in self._parameter_map.items():
+                self.prog.write_memory(region_name=region, value=value)
+
         if "pyqvm" in self.qc.name:
-            return self.qc.run(self.prog, memory_map=self._parameter_map)
+            return self.extract_samples(self.qc.run(self.prog))
 
         if self.circuit_hash is None:
             # Parametric compilation was set to False
             # Compile the program
             self._compiled_program = self.qc.compile(self.prog)
-            return self.qc.run(executable=self._compiled_program)
+            results = self.qc.run(executable=self._compiled_program)
+            return self.extract_samples(results)
```

Comment:
Careful because in this case the compiled program is not saved in the `self._compiled_program_dict` dictionary.

Reply:
That is how it was before too. It looks like the original intent was to not cache the compiled programs if parametric compilation was disabled.

The [`circuit_hash` property returns `None` if parametric compilation is disabled](https://github.com/PennyLaneAI/pennylane-forest/pull/107/files/788b3cc2f73e0bf17700882328200802d89e644b#diff-e54edc4229030c98c5b70918872608e43667b52ebfafa4f6c3d93996eea94466R267).

and the [`_compiled_program_dict` is only initialized if parametric compilation is enabled](https://github.com/PennyLaneAI/pennylane-forest/pull/107/files/788b3cc2f73e0bf17700882328200802d89e644b#diff-e54edc4229030c98c5b70918872608e43667b52ebfafa4f6c3d93996eea94466R89).

Happy to update it if that's not what we want anymore, but want to make sure we're on the same page.

Reply:
Oops my bad, you're right. Given that we are using the `circuit_hash` as dictionary keys, we can't cache it when it is None.